styx-scheduler-service/src/main/java/com/spotify/styx/StyxScheduler.java

```diff
@@ -648,27 +645,35 @@ private void setupMetrics(
               .count());
     });
 
-    workflowCache.all().forEach(workflow -> stats.registerActiveStatesMetric(
+    workflowCache.all().values().forEach(workflow -> stats.registerActiveStatesMetric(
         workflow.id(), workflowActiveStates(stateManager, workflow)));
 
     stats.registerSubmissionRateLimitMetric(submissionRateLimiter::getRate);
   }
 
-  private static Consumer<Workflow> workflowChanged(
-      WorkflowCache cache,
+  @VisibleForTesting
+  static Consumer<Workflow> workflowChanged(
       WorkflowInitializer workflowInitializer,
       Stats stats,
       StateManager stateManager,
+      Storage storage,
       BiConsumer<Optional<Workflow>, Optional<Workflow>> workflowConsumer) {
-    return (workflow) -> {
+    return workflow -> {
+      final Optional<Workflow> oldWorkflowOptional;
+      try {
+        oldWorkflowOptional = storage.workflow(workflow.id());
+      } catch (IOException e) {
+        final String message = String.format("Couldn't read workflow %s. ", workflow.id());
+        LOG.warn(message, e);
+        throw new RuntimeException(message, e);
+      }
+
       stats.registerActiveStatesMetric(
           workflow.id(), workflowActiveStates(stateManager, workflow));
 
-      final Optional<Workflow> oldWorkflowOptional = cache.workflow(workflow.id());
-
-      workflowInitializer.inspectChange(workflow);
-      cache.store(workflow);
+      workflowInitializer.inspectChange(oldWorkflowOptional, workflow);
```

Comment:
Looking up the old workflow and then mutating it seems racy to me. Use a transaction?